File: tests/pages/approvals_page.py
"""
Unified Approvals Workspace Page Object (/approvals).
"""

from playwright.sync_api import Page, expect
from tests.pages.base_page import BasePage
from tests.components.datatable import DataTableComponent


class ApprovalsPage(BasePage):
    PATH = "/approvals"

    def __init__(self, page: Page):
        super().__init__(page)
        self.table = DataTableComponent(page, name="approvals")
        self.all_items_chip = page.locator("button").filter(has_text="All Items").first
        self.leave_chip = page.locator("button").filter(has_text="Leave Requests").first
        self.reimbursement_chip = page.locator("button").filter(has_text="Reimbursements").first
        self.attendance_chip = page.locator("button").filter(has_text="Attendance Corrections").first
        self.clear_filters_button = page.locator("button").filter(has_text="Clear Filter").or_(page.locator("button").filter(has_text="Clear All Filters")).first
        self.select_all_checkbox = page.locator("[data-testid='select-all-approvals']")
        self.batch_approve_button = page.locator("[data-testid='approve-selected-btn']")

    def filter_by_module(self, module_label: str):
        """Clicks module filter chip."""
        chip = self.page.locator("button").filter(has_text=module_label).first
        expect(chip).to_be_visible()
        chip.click()
        self.page.wait_for_timeout(400)

    def approve_first_item(self):
        """Approves the first pending item in the table."""
        approve_btn = self.page.locator("button").filter(has_text="Approve").first
        expect(approve_btn).to_be_visible()
        approve_btn.click()
        self.page.wait_for_timeout(800)

File: e2e/pages/ApprovalsPage.ts
import { Page, Locator, expect } from "@playwright/test";
import { BasePage } from "./BasePage";

export class ApprovalsPage extends BasePage {
  readonly pendingCountBadge: Locator;
  readonly selectAllCheckbox: Locator;
  readonly batchApproveButton: Locator;
  readonly viewDetailButtons: Locator;
  readonly drawer: Locator;

  constructor(page: Page, baseURL?: string) {
    super(page, baseURL);
    this.pendingCountBadge = page.locator('span:has-text("Pending Action")').first();
    this.selectAllCheckbox = page.locator('[data-testid="select-all-approvals"]');
    this.batchApproveButton = page.locator('[data-testid="batch-approve-btn"], button:has-text("Batch Approve")').first();
    this.viewDetailButtons = page.locator('[data-testid="view-approval-btn"]');
    this.drawer = page.locator('[data-testid="drawer"]');
  }

  async goto(): Promise<void> {
    await this.navigate("/approvals");
  }

  async assertLoaded(): Promise<void> {
    await expect(this.page).toHaveURL(/.*\/approvals.*/);
    await expect(this.page.locator("body")).toContainText(/Approvals/i);
  }

  async filterByModule(moduleLabel: string): Promise<void> {
    const chip = this.page.locator(`button:has-text("${moduleLabel}")`).first();
    await expect(chip).toBeVisible();
    await chip.click();
    await this.page.waitForLoadState("domcontentloaded");
  }

  async openFirstDetailDrawer(): Promise<void> {
    const firstBtn = this.viewDetailButtons.first();
    await expect(firstBtn).toBeVisible();
    await firstBtn.click();
    await expect(this.drawer).toBeVisible();
  }

  async approveFirstItem(): Promise<void> {
    const btn = this.page.locator('[data-testid="approve-single-btn"], button:has-text("Approve")').first();
    await expect(btn).toBeVisible();
    await btn.click();
  }

  async rejectFirstItem(reason: string = "Policy deviation"): Promise<void> {
    const btn = this.page.locator('[data-testid="reject-single-btn"], button:has-text("Reject")').first();
    await expect(btn).toBeVisible();
    await btn.click();
    const reasonBox = this.page.locator('textarea, input[placeholder*="reason" i]').first();
    if (await reasonBox.isVisible()) {
      await reasonBox.fill(reason);
      await this.page.locator('button:has-text("Confirm Reject"), button:has-text("Reject")').last().click();
    }
  }
}
